torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
 
         # 2. Every guard must evaluate to True (but remember many guards
         #    like s0 == s1*2 because trivial due to simplification)
-        for g, tb in self.guards:
-            if self._maybe_evaluate_static(g) is not None:
+        def simplify_until_no_change(expr: sympy.Expr) -> sympy.Expr:
+            new_expr = self.simplify(expr)
+            while new_expr != expr:
+                expr = new_expr
+                new_expr = self.simplify(expr)
+            return expr
+
+        simplified_guards = [(simplify_until_no_change(g), tb) for g, tb in self.guards]
+        var_to_guards = dict()
+        var_to_range = self.var_to_range.copy()
+        unhandled_guards = []
+
+        for g, tb in simplified_guards:
+            if self._maybe_evaluate_static_impl(g, var_to_range=var_to_range) is not None:
+                continue
+
+            if not isinstance(g, sympy.Rel):
+                unhandled_guards.append((g, tb))
+                continue
+
+            symbol = g.lhs
+
+            if not isinstance(g.lhs, sympy.Symbol):
+                unhandled_guards.append((g, tb))
                 continue
-            g = self.simplify(g)
+
+            assert symbol in self.var_to_range
+
+            # Consider actually issuing the guard if it is used for refining
+            # the value range of a symbol. Otherwise, it is discarded.
+
+            vr = var_to_range[symbol]
+            lower, upper = vr.lower, vr.upper
+
+            rhs_vr = sympy_interp(ValueRangeAnalysis, self.var_to_range, g.rhs)
+            lower_guard, upper_guard = var_to_guards.get(symbol, (None, None))
+
+            if lower < rhs_vr.lower and isinstance(g, (sympy.Eq, sympy.Ge, sympy.Gt)):
+                lower = rhs_vr.lower + (1 * isinstance(g, sympy.Gt))
+                lower_guard = (g, tb)
+            if upper < rhs_vr.upper and isinstance(g, (sympy.Eq, sympy.Le, sympy.Lt)):
+                upper = rhs_vr.upper - (1 * isinstance(g, sympy.Lt))
+                upper_guard = (g, tb)
+
+            # If the value range is still the same, ignore the guard.
```

Comment:
This seems wrong.

I can have a guard which doesn't improve my range at all, but is still required; e.g., a guard that x is even doesn't refine range, but you can't drop it.

Reply:
You are right. I guess the safe choice here is to let `_maybe_evaluate_static` decide.